### PR TITLE
Bugfix/akka 27647

### PR DESCRIPTION
**Pull request by sebarys**
## References
https://github.com/akka/akka/issues/27647

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 1]
```diff
@@ -359,8 +359,9 @@ private[akka] class Shard(
     case None =>
       log.debug("HandOff shard [{}]", shardId)
 
-      if (state.entities.nonEmpty) {
+      if (idByRef.nonEmpty) {
         val entityHandOffTimeout = (settings.tuningParameters.handOffTimeout - 5.seconds).max(1.seconds)
+        log.debug("Starting HandOffStopper for shard {} to terminate following entities: {}", shardId, idByRef.keySet)
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: I think that could risk being a very large number of ids, maybe log the number of entities instead to not do a denial of service on the logfile?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 2]
```diff
@@ -397,10 +397,12 @@ object ShardRegion {
     def receive = {
       case ReceiveTimeout =>
         log.warning(
-          "HandOffStopMessage[{}] is not handled by some of the entities of the `{}` shard, " +
-          "stopping the remaining entities.",
+          "HandOffStopMessage[{}] is not handled by some of the entities of the `{}` shard after {}, " +
+          "stopping the remaining entities: {}",
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: Same risk here

**Review by johanandren — COMMENTED** [review]
Fix looks good, not sure about logging.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 1 above)
Comment (sebarys) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: good point, I've changed to log number of entities

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala (hunk 2 above)
Comment (sebarys) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: changed to log number of entites

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

**Review by patriknw — APPROVED** [review]
LGTM, somewhat different conventions for the logging than what we normally have but that was probably already the case. I'll look into that.

**Pull request merged**